Use Nginx as reverse proxy

little-dude opened the pull request:
I found your image thanks to https://dantehranian.wordpress.com/2014/09/03/a-local-caching-proxy-for-pypi-python-org-via-docker/ 

I'd like to use NGINX to do https. This PR does not introduce HTTPS yet, but it adds nginx as reverse proxy, which is the first step.

There are some other smaller changes (see the commit message) which are not directly related, so let me know if you want me to do several PRs.

From a user point of view those changes should be transparent. The behaviour is unchanged, but hunder the hood, the image uses nginx as reverse proxy.

Note that despite the fact that everything is working, I'm seeing errors in nginx logs sometimes :

```
2015/04/09 15:45:05 [error] 78#0: *8 connect() failed (111: Connection refused) while connecting to upstream, client: 172.17.42.1, server: localhost, request: "GET /root/dev HTTP/1.1", upstream: "http://[::1]:3142/root/dev", host: "localhost:3141"
```

I'm not familiar with nginx so if you have any idea to fix this...

Commit:
Use Nginx as reverse proxy

Use Nginx to forward devpi-client requests to devpi-server.

Changes introduced :
--------------------

Components :
    - Nginx is installed

Behavior :
    - Nginx listens on 0.0.0.0:3141. Ip and port are configurable via
      NGINX_IP and NGINX_PORT environment variables.
    - if NGINX_ENABLETLS is true, nginx also listens on port 3142 for
      HTTPS requests. TLS is configurable via multiple environment
      variables that are documented in the example environment file.
    - devpi-server listens on 127.0.0.1:3143. Ip and port are
      configurable via DEVPI_SERVERIP and DEVPI_SERVERPORT environment
      variables.

Files :
    - devpi-nginx.env : example of environment file to use with this
      container. Environment are documented there.
    - nginx-tpl.conf : nginx configuration template
    - devpi.crt and devpi.key : dumb certificate added to the container
      and used by default when https is enabled. Users are encouraged
      not to use it. The README explains how to use custom certs.

References :
------------

    - https://github.com/nginxinc/docker-nginx the official nginx docker
      image repo
    - https://dantehranian.wordpress.com/2014/09/03/a-local-caching-proxy-for-pypi-python-org-via-docker/
      blog post about nginx+docker
    - doc.devpi.net/latest/userman/index.html devpi documentation

tehranian commented:
Couldn't you run the nginx proxy on the host OS and have it handle all of the HTTPS bits, then proxy to the devpi app in the backend with regular HTTP?

In other words, does the nginx service really need to be within the container?

little-dude commented:
Yes, what you say is tperfeclty doable.
The reasons I'm doing it like this are : 

- it transparent for the user : the container behave like it was behaving before my changes.
- there is no need to think about how to configure nginx, or how to link your containers together, everything is done in this one.
- to me (but this is personnal opinion) pypi over https is one service, not two, and thus I'd like to be able to have this service provided by a single image.

dangra commented:
Thanks for looking into this, I am generally +1 to merge but see my comments.

IMHO https is not a reason to include nginx into the container but serving static files using nginx it is. If you managed to send a pull request later with a non intrusive way to enable TLS (disabled by default), I'll merge it.

dangra commented on Dockerfile:
```diff
@@ -1,11 +1,27 @@
+        "devpi-client" \
+        "requests" \
+        "devpi-ldap" \
+        "devpi-web"
```
Pinning devpi versions is important to have reproducible builds, so please restore the original devpi package versions and we can look at upgrading them in another PR.

little-dude replied:
Makes sense, I'll restore this.

dangra commented on Dockerfile:
```diff
@@ -1,11 +1,27 @@
+        "devpi-web"
+
+# By default expose a data volume and port 3141
+VOLUME /devpidata
```
Do not change volume names, the new image must be backwards compatible.

little-dude replied:
:+1:

dangra commented on run.sh:
```diff
@@ -1,15 +1,71 @@
+if [[ ! -f ${DEVPI_SERVERDIR}/.serverversion ]]; then
+  devpi use http://${DEVPI_SERVERIP}:${DEVPI_SERVERPORT}
+  devpi login ${DEVPI_CLIENTNAME} --password="${DEVPI_CLIENTPWD}"
+  devpi user -m root password="${DEVPI_CLIENTPWD}"
```
Although DEVPI_CLIENTPWD seems more accurate, the old name DEVPI_PASSWORD is still mentioned in README and we should retain backwards compatibility.

dangra commented on run.sh:
```diff
@@ -1,15 +1,71 @@
+[[ -z ${DEVPI_CLIENTDIR+xxx} ]]   && export DEVPI_CLIENTDIR=$DEFAULT_DEVPI_CLIENTDIR
+[[ -z ${DEVPI_CLIENTNAME+xxx} ]]  && export DEVPI_CLIENTNAME=$DEFAULT_DEVPI_CLIENTNAME
+[[ -z ${DEVPI_CLIENTPWD+xxx} ]]   && export DEVPI_CLIENTPWD=$DEFAULT_DEVPI_CLIENTPWD
+[[ -z ${DEVPI_LDAPCONFILE+xxx} ]] && export DEVPI_LDAPCONFILE=$DEFAULT_DEVPI_LDAPCONFILE
```
I think all of this can be reduced to:

```
: ${DEVPI_SERVERDIR:=/devpidata}
: ${NGINX_IP:=0.0.0.0}
: ${NGINX_PORT:=3141}
...
export DEVPI_SERVERDIR NGINX_IP NGINX_PORT ...
```

little-dude replied:
Woa, I didn't know this syntax, thanks!

dangra commented on run.sh:
```diff
@@ -1,15 +1,71 @@
+sed -i "s/{{DEVPI_SERVERPORT}}/${DEVPI_SERVERPORT}/g" $nginx_conf_file 
+# DEVPI_SERVERDIR contains / so it needs to be escaped
+# See http://stackoverflow.com/questions/407523/escape-a-string-for-a-sed-replace-pattern
+sed -i "s/{{DEVPI_SERVERDIR}}/$(echo ${DEVPI_SERVERDIR} | sed -e 's/[\/&]/\\&/g')/g" $nginx_conf_file 
```
I have used nginx as reverse proxy for other services, and found more practical to ship the whole nginx.conf with `http` and `server` sections in a single file than reusing default configuration files and _/etc/nginx/sites-enable_ method.

for example this is the reverse proxy for docker registry:
```
daemon off;    
user www-data;    
worker_processes 4;    
pid /run/nginx.pid;    
    
events {    
  worker_connections 768;    
}    
    
http {    
  tcp_nopush on;    
  tcp_nodelay on;    
  keepalive_timeout 65;    
  types_hash_max_size 2048;    
  access_log /var/log/nginx/access.log;    
  error_log /var/log/nginx/error.log;    
  gzip off;    
    
  upstream docker-registry {    
    server ${REGISTRY_PORT_5000_TCP_ADDR}:${REGISTRY_PORT_5000_TCP_PORT};    
  }    
    
  server {    
    listen 443;    
    server_name i.mydomain.com;    
    ssl on;    
    ssl_certificate /opt/nginx/cert.crt;    
    ssl_certificate_key /opt/nginx/cert.key;    
    
    # disable any limits to avoid HTTP 413 for large image uploads    
    client_max_body_size 0;    
    
    proxy_set_header Host           $host;    
    proxy_set_header X-Real-IP      $remote_addr;    
    proxy_set_header Authorization  "";    
    
    location /ca.crt {    
        alias /opt/nginx/ca.crt;    
    }    
    
    location / {    
      auth_basic "Registry";    
      auth_basic_user_file /opt/nginx/htpasswd;    
      proxy_pass  http://docker-registry;    
    }    
    
    location /v1/_ping {    
      proxy_pass  http://docker-registry;    
    }    
    
  }    
}    
```

I store it as nginx-tpl.conf and use a shell script to generate the actual nginx.conf file:

```bash
#!/bin/bash    
env \    
  |sed 's/[\%]/\\&/g;s/\([^=]*\)=\(.*\)/s%${\1}%\2%/' \    
  |sed -f- /opt/nginx/nginx.conf \    
  |tee /tmp/nginx.conf    
exec nginx -c /tmp/nginx.conf "$@"
```

as you see (or not) it is generating sed rules for every environment variable and using it to render the nginx conf template. Above script is docker image command.

dangra replied:
btw, I have tested my sed rules with slashes in envvar values and it worked fine.

little-dude commented:
> Thanks for looking into this, I am generally +1 to merge but see my comments.
Thanks for reviewing and commenting. I'll repush soon.

> IMHO https is not a reason to include nginx into the container but serving static files using nginx it is. If you managed to send a pull request later with a non intrusive way to enable TLS (disabled by default), I'll merge it.

I was planning on adding a `HTTPS` env variable. If set to true, both http and https would be enabled (on ports 3141 and 3142 respectively). The image would be shipped with a default self-signed certificate, but I'd specify in the README that it's recommended to use your own CA. Does that sound good?

dangra commented:
> I was planning on adding a HTTPS env variable. If set to true, both http and https would be enabled (on ports 3141 and 3142 respectively). The image would be shipped with a default self-signed certificate, but I'd specify in the README that it's recommended to use your own CA. Does that sound good?

:ok_hand:

little-dude commented:
@dangra I re-pushed taking your comments into account, and squashed everything to make reviewing easier. I also added the HTTPS config to the nginx template. Thanks for the `env | sed` command, it's pretty awesome.

Please don't merge, since for the moment, https does not work and i have no idea why.

dangra commented on run.sh:
```diff
@@ -1,15 +1,57 @@
+NGINX_TLSPORT=${NGINX_TLSPORT:-3142}
+NGINX_TLSPATH=${NGINX_TLSPATH:-/opt/nginx/certs/default}
+NGINX_TLSCERT=${NGINX_TLSCERT:-devpi.crt}
+NGINX_TLSKEY=${NGINX_KEY:-devpi.key}
```
why did you use `VAR = ${VAR:-value}` syntax instead of `: ${VAR:=value}` ?

little-dude replied:
Oh, just a mistake. I didn't know this syntax, so I googled it and copy/paste from stackoverflow where it was slightly different. Will change for `: ${VAR:=value}`

dangra commented on run.sh:
```diff
@@ -1,15 +1,57 @@
+    # Configure the default client if necessary
+    if [[ ! -f ${DEVPI_SERVERDIR}/.serverversion ]]; then
+      devpi use http://${DEVPI_SERVERIP}:${DEVPI_SERVERPORT}
+      devpi login root --password="${DEVPI_CLIENTPASSWORD}"
```
Default root password is an empty string so first login can't use _DEVPI_CLIENTPASSWORD_

little-dude replied:
Indeed :+1:

Commit:
use concise syntaxe to set var with default values

Use:
: ${VAR:=default_value}

instead of:
VAR=${VAR:-default_value}

Commit:
Use empty pwd for root for first login

When start devpi-server, the root password is empty, so we can't login
with the password provided by the environment variable DEVPI_PASSWORD
passed when starting the container.

Commit:
use default values in the example environment file

default values are NGINX_ENABLETLS=false, and
NGINX_TLSPATH=/opt/nginx/certs/default

little-dude commented on run.sh:
```diff
@@ -1,15 +1,57 @@
+    # Escape all \ & and / for every variable of the current env, and replace all
+    # ${var} occurences in nginx conf template file with the corresponding variable
+    # value
+    env | sed 's/[\%]/\\&/g;s/\([^=]*\)=\(.*\)/s%${\1}%\2%/' | sed -f- ${nginx_template_file} | tee /tmp/nginx.conf    
```
Seeing a very weird issue here : the `tee` is sometimes executed *before* the `sed -f-` : 

```
+ env
+ sed 's/[\%]/\\&/g;s/\([^=]*\)=\(.*\)/s%${\1}%\2%/'
+ tee /tmp/nginx.conf
+ sed -f- /opt/nginx/nginx-tpl.conf
```

wth??

dangra replied:
does it matters? Bash should pipe correct stdin/stdout file descriptors anyway

little-dude replied:
I thought I matters because substitutions were not happening in the nginx template. But this is actually because with the new syntax `: ${var:=default_val}`, I need to export the variables.

dangra replied:
ok, right.

2015-04-13 13:45 GMT-03:00 little-dude <notifications@github.com>:

> In run.sh
> <https://github.com/scrapinghub/docker-devpi/pull/2#discussion_r28255222>:
>
> > +      devpi index -y -c public pypi_whitelist='*'
> > +    fi
> > +}
> > +
> > +function start_nginx () {
> > +    nginx_template_file=/opt/nginx/nginx-tpl.conf
> > +
> > +    # Uncomment lines that enable TLS in template file
> > +    if [ ${NGINX_ENABLETLS} = true ]; then
> > +        sed -i "s/#{uncomment_if_tls_enabled}\s*//g" ${nginx_template_file}
> > +    fi
> > +
> > +    # Escape all \ & and / for every variable of the current env, and replace all
> > +    # ${var} occurences in nginx conf template file with the corresponding variable
> > +    # value
> > +    env | sed 's/[\%]/\\&/g;s/\([^=]*\)=\(.*\)/s%${\1}%\2%/' | sed -f- ${nginx_template_file} | tee /tmp/nginx.conf
>
> I thought I matters because substitutions were not happening in the nginx
> template. But this is actually because with the new syntax :
> ${var:=default_val}, I need to export the variables.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scrapinghub/docker-devpi/pull/2/files#r28255222>.
>

Commit:
fix variable scope in /run.sh

fix 7856cb49337ee88c36377c26e77fd6cb82fcee51

After defining variable with
: ${var:=val}
we need to export the variables, which was not necessary when using
var=${var:-val}

no idea why, though.

dangra commented on Dockerfile:
```diff
@@ -1,11 +1,35 @@
+EXPOSE 3142
+
+# devpi-server will listen on 127.0.0.1:3143 by default
+EXPOSE 3143
```
you can group exposed ports into a single EXPOSE directive, see https://docs.docker.com/reference/builder/#expose

It will generate one image layer instead of 3.

dangra commented on devpi-nginx.env:
```diff
@@ -0,0 +1,41 @@
+DEVPI_CLIENTDIR=/tmp/devpi-client
+
+# For future use
+DEVPI_LDAPCONFILE=
```
Please remove all references to LDAPCONFILE, if we eventually add it will come in its own pull request.

dangra commented on Dockerfile:
```diff
@@ -1,11 +1,35 @@
+
+# For TLS, build the container with a cert and a key
+ADD devpi.crt /opt/nginx/default/
+ADD devpi.key /opt/nginx/default/
```
Let's put all nginx related files in a subdirectory named `nginx` and a single _ADD_ directive: `ADD nginx /opt/nginx`, no need for _mkdir_ and one layer instead of 4.

no default or custom subdirs, just _/op/nginx/certs/_ and let's enable TLS by default using the selfsigned certs you generated.

Users can bind mount _/opt/nginx/certs_ with their own certificates if they want.

dangra replied:
`/opt/nginx/` directory structure will be:

```
/opt/nginx/nginx-tpl.conf
/opt/nginx/certs/devpi.key
/opt/certs/devpi.crt
```

dangra commented on run.sh:
```diff
@@ -1,15 +1,58 @@
+: ${DEVPI_SERVERPORT:=3143}
+: ${DEVPI_SERVERDIR:=/mnt}
+: ${DEVPI_CLIENTDIR:=/tmp/devpi-client}
+: ${DEVPI_CLIENTPASSWORD:=""}
```
Remember the image should be backwards compatible for previously used name `DEVPI_PASSWORD` (still mentioned in README)

just replace this line by `: ${DEVPI_CLIENTPASSWORD:=$DEVPI_PASSWORD}`

dangra commented on run.sh:
```diff
@@ -1,15 +1,58 @@
+: ${NGINX_TLSPATH:=/opt/nginx/certs/default}
+: ${NGINX_TLSCERT:=devpi.crt}
+: ${NGINX_KEY:=devpi.key}
+export DEVPI_SERVERIP DEVPI_SERVERPORT DEVPI_SERVERDIR DEVPI_CLIENTDIR DEVPI_CLIENTPASSWORD DEVPI_LDAPCONFILE NGINX_IP NGINX_PORT NGINX_ENABLETLS NGINX_TLSPORT NGINX_TLSPATH NGINX_TLSCERT NGINX_KEY
```
another nice trick is to export everything prefixed by `NGINX_` or `DEVPI_`

`export ${!DEVPI_@} ${!NGINX_@}`

dangra commented on run.sh:
```diff
@@ -1,15 +1,58 @@
+}
+
+function start_nginx () {
+    nginx_template_file=/opt/nginx/nginx-tpl.conf
```
It could be good to make nginx_template_file a setting variable with default value `/opt/nginx/nginx-tpl.conf`

bubenkoff commented:
wouldn't make more sense to keep the container running only one service
you can always add nginx as proxy running in another container and use container linking to connect it with devpi.
Here's the example how i do it when running in amazon elastic beanstalk
https://github.com/InnovativeTravel/devpi-beanstalk
